pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,276 @@
+
+func (s *downwardMetricsServer) start(stop chan struct{}) error {
+	conn, err := connect(s.virtioSerialSocket, s.connectAttempts)
+	if err != nil {
+		return err
+	}
+
+	go s.serve(conn, stop)
+	return nil
+}
+
+func (s *downwardMetricsServer) serve(conn net.Conn, stop chan struct{}) {
+	defer func(conn net.Conn) {
+		err := conn.Close()
+		if err != nil {
+			log.Log.Reason(err).Warning("virtio-serial failed to close the connection")
+		}
+	}(conn)
+
+	type reqResult struct {
+		request string
+		err     error
+	}
+	newRequest := make(chan reqResult)
+	defer close(newRequest)
+	reader := textproto.NewReader(bufio.NewReader(conn))
+
+	for {
+		go func() {
+			// Let's ignore the possible panic for sending on a closed channel
+			defer func() { _ = recover() }()
+
+			requestLine, err := waitForRequest(reader)
+			if err != nil && errors.Is(err, net.ErrClosed) {
+				// Closing the connection signals that the server should stop
+				return
+			}
+
+			newRequest <- reqResult{request: requestLine, err: err}
+		}()
+
+		select {
+		case res := <-newRequest:
+			if res.err != nil {
+				if errors.Is(res.err, io.EOF) {
+					// qemu closed the connection
+					return
+				}
+				log.Log.Reason(res.err).Warning("virtio-serial socket read failed")
+				continue
+			}
+
+			response, err := s.handleRequest(res.request)
+			if err != nil {
+				log.Log.Reason(err).Error("failed to process the request")
+				replyError(conn)
+				continue
+			}
+
+			err = reply(conn, response)
+			if err != nil {
+				log.Log.Reason(err).Error("failed to send the metrics")
+			}
+		case <-stop:
+			return
+		}
+	}
+}
+
+func (s *downwardMetricsServer) handleRequest(requestLine string) ([]byte, error) {
+	err := parseRequest(requestLine)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := s.getXmlMetrics()
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func (s *downwardMetricsServer) getXmlMetrics() ([]byte, error) {
+	var xmlMetrics []byte
+	metrics, err := s.report()
+	if err != nil {
+		xmlMetrics = emptyMetrics
+		log.Log.Reason(err).Error("failed to collect the metrics")
+	} else {
+		xmlMetrics, err = xml.MarshalIndent(metrics, "", "  ")
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode metrics: %v", err)
+		}
+	}
+
+	// `vm-dump-metrics` expects `\n\n` as a termination symbol
+	xmlMetrics = append(xmlMetrics, '\n', '\n')
+	return xmlMetrics, nil
+}
+
+func connect(socketPath string, attempts uint) (net.Conn, error) {
+	var conn net.Conn
+	var err error
+
+	multiplier := 1
+	for i := uint(0); i < attempts; i++ {
+		conn, err = net.Dial("unix", socketPath)
+		if err == nil {
+			break
+		}
+
+		// It is only tried again in case the socket doesn't exist or not one is
+		// listening on the other end yet
+		if !(errors.Is(err, syscall.ECONNREFUSED) || errors.Is(err, syscall.ENOENT)) {
+			break
+		}
+
+		if i == attempts {
+			break // it was the last attempt
+		}
+
+		backoff := time.Duration(multiplier) * time.Second
```

Comment:
Could a static backoff time be enough?

Reply:
I think so, we already run the server after qemu started so (in theory) we should only worry about slow/busy disks

Reply:
I don't really think is unnecessary, true it would be simple the whole `connect()` will be just `conn, err = net.Dial("unix", socketPath)`, but I rather to have some way to wait a bit if the socket is not ready than just fail unconditionally.

Reply:
I was just talking about the multiplier in the backoff time. My question was if a static backoff time would be enough?

Reply:
(ah ok I misunderstood it, sorry)
Probably (with timeouts you never know for sure:) ), I can leave waiting time of 1s, and increase the number of maxConnAttempts. The only benefit of using a backoff time is to not contribute to saturate an already saturated system., WDYT?

Reply:
Let's leave it like this.